### PR TITLE
added head command, which returns a VDS with the N first elements fro…

### PR DESCRIPTION
**Pull request by lfrancioli**
…m the first partition(s) a VDS
- added head driver
- added HeadRDD in Utils to create a new RDD keeping only N elements from its parent
- added head function in VSM

### CODE REVIEW

File: src/main/scala/org/broadinstitute/hail/Utils.scala [hunk 1]
```diff
@@ -324,6 +324,53 @@ class RichSparkContext(val sc: SparkContext) extends AnyVal {
...
+
+  private var nLastPart = 0
+  private var lastPart = 0
+
+  def getPartitions: Array[Partition] = {
+    val buf = new mutable.ArrayBuffer[Int]
+
+    buf ++= prev.sparkContext.runJob(prev, (it: Iterator[T]) => it.length, Seq(0))
+    if (exact || buf.length == 0) {
+      while (buf.sum < num && buf.length < prev.partitions.length) {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/Utils.scala]: Recalling `sum` is also quadratic.  I'd keep a running sum.

File: src/main/scala/org/broadinstitute/hail/Utils.scala [hunk 2]
```diff
@@ -324,6 +324,53 @@ class RichSparkContext(val sc: SparkContext) extends AnyVal {
...
+        buf ++= prev.sparkContext.runJob(prev, (it: Iterator[T]) => it.length,
+          buf.length to
+            buf.length +
+              (if (buf.sum > 0) math.min(((num - buf.sum) / (buf.sum / buf.length).toDouble).ceil.toInt, prev.partitions.length)
+              else 1)
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/Utils.scala]: If the first bunch of partitions are empty, or very small compared to what's to come, this isn't going to perform very well.  A few ideas:

 - I'd start with 4 partitions, not 1.  Most systems we run on will have at least 4 executors (even laptops), so it should essentially be free.

 - You might increase exponentially by 4s until you reach some fraction of the target (10%?)  If you ask for 10k and there is 1 element in the first partition, you're going to compute the entire thing even if the second partition has >10k entries.

 - Use the >10% as an estimate in no-exact mode, and a first guess in exact mode.

What do you think?

File: src/main/scala/org/broadinstitute/hail/Utils.scala [hunk 3]
```diff
@@ -324,6 +324,53 @@ class RichSparkContext(val sc: SparkContext) extends AnyVal {
...
+        if (n + buf(i) > num) {
+          lastPart = i
+          nLastPart = num - n
+        } else {
+          f(n + buf(i), i + 1)
+        }
+      }
+
+      f(0, 0)
+      firstParent.partitions
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/Utils.scala]: I think you can do the filter, and compute the partitions along with setting lastPart and nLastPart all together in `f`.  It reads awkwardly to go over partitions twice right together using two different idioms (recursion vs combinators).

File: src/main/scala/org/broadinstitute/hail/Utils.scala [hunk 4]
```diff
@@ -324,6 +324,53 @@ class RichSparkContext(val sc: SparkContext) extends AnyVal {
...
+        buf ++= prev.sparkContext.runJob(prev, (it: Iterator[T]) => it.length,
+          buf.length to
+            buf.length +
+              (if (buf.sum > 0) math.min(((num - buf.sum) / (buf.sum / buf.length).toDouble).ceil.toInt, prev.partitions.length)
+              else 1)
+        )
+      }
+
+      def f(n: Int, i: Int): Unit = {
+        if (n + buf(i) > num) {
```
Comment (cseed) [inline src/main/scala/org/broadinstitute/hail/Utils.scala]: This should be `>=`, right?

**Comment by cseed** [comment]
OK, a few more comments.  I think it'll be ready to go after this!

**Comment by tpoterba** [comment]
Let's get this in.  Can you rebase?

**Comment by cseed** [comment]
I am part way through implementing my suggestions.  I'll take responsibility for this.

**Comment by lfrancioli** [comment]
Of course. Cotton also had a few comments, I will integrate too.

**Comment by cseed** [comment]
I'm going to PR an alternate version of this (some day).

**Pull request closed without merging**